From dc285cc3fe0d7b3c91eb46ce218fddfe580eaf17 Mon Sep 17 00:00:00 2001
From: Jan Dominiak <janx.dominiak@intel.com>
Date: Wed, 22 Apr 2020 07:57:05 +0200
Subject: [PATCH] PONRTSYS-6060: Add missing LAN Ethernet PM counters to
 ethtool

Mib_counters implementation extended to support new counters:
* RxOverflowError
* RxGoodPausePkts
* TxPauseCount
* TxExcessCollCount
* TxLateCollCount
* TxSingleCollCount
* TxMultCollCount
---
 drivers/net/datapath/dpm/Makefile                  |   1 -
 drivers/net/datapath/dpm/datapath.h                |  53 ---------
 drivers/net/datapath/dpm/datapath_api.c            |  59 ++++++----
 drivers/net/datapath/dpm/datapath_stats.c          |  28 -----
 drivers/net/datapath/dpm/datapath_stats.h          |  30 -----
 drivers/net/datapath/dpm/gswip31/datapath_misc.c   | 111 ------------------
 drivers/net/datapath/dpm/gswip31/datapath_misc.h   |  15 ---
 .../net/datapath/mib_counters/dp_mib_counters.c    | 128 +++++++++++++++++++--
 include/net/datapath_api.h                         |  22 +++-
 9 files changed, 177 insertions(+), 270 deletions(-)

diff --git a/drivers/net/datapath/dpm/Makefile b/drivers/net/datapath/dpm/Makefile
index c27441c4d432..c6ac4cf4cbf4 100644
--- a/drivers/net/datapath/dpm/Makefile
+++ b/drivers/net/datapath/dpm/Makefile
@@ -1,5 +1,4 @@
 obj-$(CONFIG_INTEL_DATAPATH) = datapath_api.o datapath_proc_api.o datapath_proc.o  datapath_misc.o datapath_notifier.o datapath_logical_dev.o datapath_instance.o datapath_soc.o datapath_qos.o datapath_proc_qos.o datapath_tx.o datapath_cqm.o
-obj-$(CONFIG_INTEL_DATAPATH) += datapath_stats.o
 
 ifneq ($(CONFIG_INTEL_DATAPATH_HAL_GSWIP32),)
 obj-$(CONFIG_INTEL_DATAPATH) += gswip32/
diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 6e8620b6ec0f..80bbd00bc2f4 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -364,59 +364,6 @@ enum QOS_FLAG {
 	QOS_PORT_CFG_SET, /* set qos port config info */
 };
 
-/* Ethtool strings and counter names (based on GSW_RMON_Port_cnt_t) */
-#define DP_ETHTOOL_STAT_REG_DECLARE	\
-	_DP_ETHTOOL(RxExtendedVlanDiscardPkts, nRxExtendedVlanDiscardPkts) \
-	_DP_ETHTOOL(MtuExceedDiscardPkts, nMtuExceedDiscardPkts) \
-	_DP_ETHTOOL(TxUnderSizeGoodPkts, nTxUnderSizeGoodPkts) \
-	_DP_ETHTOOL(TxOversizeGoodPkts, nTxOversizeGoodPkts) \
-	_DP_ETHTOOL(RxGoodPkts, nRxGoodPkts) \
-	_DP_ETHTOOL(RxUnicastPkts, nRxUnicastPkts) \
-	_DP_ETHTOOL(RxBroadcastPkts, nRxBroadcastPkts) \
-	_DP_ETHTOOL(RxMulticastPkts, nRxMulticastPkts) \
-	_DP_ETHTOOL(RxFCSErrorPkts, nRxFCSErrorPkts) \
-	_DP_ETHTOOL(RxUnderSizeGoodPkts, nRxUnderSizeGoodPkts) \
-	_DP_ETHTOOL(RxOversizeGoodPkts, nRxOversizeGoodPkts) \
-	_DP_ETHTOOL(RxUnderSizeErrorPkts, nRxUnderSizeErrorPkts) \
-	_DP_ETHTOOL(RxGoodPausePkts, nRxGoodPausePkts) \
-	_DP_ETHTOOL(RxOversizeErrorPkts, nRxOversizeErrorPkts) \
-	_DP_ETHTOOL(RxAlignErrorPkts, nRxAlignErrorPkts) \
-	_DP_ETHTOOL(RxFilteredPkts, nRxFilteredPkts) \
-	_DP_ETHTOOL(Rx64BytePkts, nRx64BytePkts) \
-	_DP_ETHTOOL(Rx127BytePkts, nRx127BytePkts) \
-	_DP_ETHTOOL(Rx255BytePkts, nRx255BytePkts) \
-	_DP_ETHTOOL(Rx511BytePkts, nRx511BytePkts) \
-	_DP_ETHTOOL(Rx1023BytePkts, nRx1023BytePkts) \
-	_DP_ETHTOOL(RxMaxBytePkts, nRxMaxBytePkts) \
-	_DP_ETHTOOL(TxGoodPkts, nTxGoodPkts) \
-	_DP_ETHTOOL(TxUnicastPkts, nTxUnicastPkts) \
-	_DP_ETHTOOL(TxBroadcastPkts, nTxBroadcastPkts) \
-	_DP_ETHTOOL(TxMulticastPkts, nTxMulticastPkts) \
-	_DP_ETHTOOL(TxSingleCollCount, nTxSingleCollCount) \
-	_DP_ETHTOOL(TxMultCollCount, nTxMultCollCount) \
-	_DP_ETHTOOL(TxLateCollCount, nTxLateCollCount) \
-	_DP_ETHTOOL(TxExcessCollCount, nTxExcessCollCount) \
-	_DP_ETHTOOL(TxCollCount, nTxCollCount) \
-	_DP_ETHTOOL(TxPauseCount, nTxPauseCount) \
-	_DP_ETHTOOL(Tx64BytePkts, nTx64BytePkts) \
-	_DP_ETHTOOL(Tx127BytePkts, nTx127BytePkts) \
-	_DP_ETHTOOL(Tx255BytePkts, nTx255BytePkts) \
-	_DP_ETHTOOL(Tx511BytePkts, nTx511BytePkts) \
-	_DP_ETHTOOL(Tx1023BytePkts, nTx1023BytePkts) \
-	_DP_ETHTOOL(TxMaxBytePkts, nTxMaxBytePkts) \
-	_DP_ETHTOOL(TxDroppedPkts, nTxDroppedPkts) \
-	_DP_ETHTOOL(TxAcmDroppedPkts, nTxAcmDroppedPkts) \
-	_DP_ETHTOOL(RxDroppedPkts, nRxDroppedPkts) \
-	_DP_ETHTOOL(RxGoodBytes, nRxGoodBytes) \
-	_DP_ETHTOOL(RxBadBytes, nRxBadBytes) \
-	_DP_ETHTOOL(TxGoodBytes, nTxGoodBytes)
-
-enum DP_ETHTOOL_STAT {
-#define _DP_ETHTOOL(x, y)	DP_E_STAT_ ## x,
-DP_ETHTOOL_STAT_REG_DECLARE
-#undef _DP_ETHTOOL
-};
-
 /* Tx counters serviced by CTP bypass registers */
 #define DP_ETHTOOL_STAT_REG_TX_DECLARE \
 	_DP_ETHTOOL_TX(nTxUnderSizeGoodPkts) \
diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 4984f2d81a10..183052c234ae 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -107,12 +107,33 @@ void (*dp_dev_get_ethtool_stats_fn)(struct net_device *dev,
 EXPORT_SYMBOL(dp_dev_get_ethtool_stats_fn);
 
 void dp_set_ethtool_stats_fn(int inst, void (*cb)(struct net_device *dev,
-			struct ethtool_stats *stats, u64 *data))
+			     struct ethtool_stats *stats, u64 *data))
 {
 	dp_dev_get_ethtool_stats_fn = cb;
 }
 EXPORT_SYMBOL(dp_set_ethtool_stats_fn);
 
+int (*dp_get_dev_stat_strings_count_fn)(struct net_device *dev) = NULL;
+EXPORT_SYMBOL(dp_get_dev_stat_strings_count_fn);
+
+void dp_set_ethtool_stats_strings_cnt_fn(int inst,
+					 int (*cb)(struct net_device *dev))
+{
+	dp_get_dev_stat_strings_count_fn = cb;
+}
+EXPORT_SYMBOL(dp_set_ethtool_stats_strings_cnt_fn);
+
+void (*dp_get_dev_ss_stat_strings_fn)(struct net_device *dev,
+				      u8 *data) = NULL;
+EXPORT_SYMBOL(dp_get_dev_ss_stat_strings_fn);
+
+void dp_set_ethtool_stats_strings_fn(int inst, void (*cb)(struct net_device *dev,
+				     u8 *data))
+{
+	dp_get_dev_ss_stat_strings_fn = cb;
+}
+EXPORT_SYMBOL(dp_set_ethtool_stats_strings_fn);
+
 int dp_register_ops(int inst, enum DP_OPS_TYPE type, void *ops)
 {
 	if (is_invalid_inst(inst) || type >= DP_OPS_CNT) {
@@ -2689,34 +2710,33 @@ int bp_pmapper_dev_get(int inst, struct net_device *dev)
 /* ethtool statistics support */
 void dp_net_dev_get_ss_stat_strings(struct net_device *dev, u8 *data)
 {
-	dp_subif_t subif = {0};
-	struct inst_info *pinfo = &dp_port_prop[subif.inst].info;
-
-	if (!pinfo->dp_net_dev_get_ss_stat_strings)
-		return;
+	/* This is for the Mib counter wraparound module */
+	if (dp_get_dev_ss_stat_strings_fn) {
+		DP_DEBUG(DP_DBG_FLAG_MIB,
+			 "MIB call dp_get_dev_ss_stat_strings_fn callback\n");
+		dp_get_dev_ss_stat_strings_fn(dev, data);
+	}
 
-	pinfo->dp_net_dev_get_ss_stat_strings(dev, data);
+	return;
 }
 EXPORT_SYMBOL(dp_net_dev_get_ss_stat_strings);
 
 int dp_net_dev_get_ss_stat_strings_count(struct net_device *dev)
 {
-	dp_subif_t subif = {0};
-	struct inst_info *pinfo = &dp_port_prop[subif.inst].info;
-
-	if (!pinfo->dp_net_dev_get_ss_stat_strings_count)
-		return -EOPNOTSUPP;
+	/* This is for the Mib counter wraparound module */
+	if (dp_get_dev_stat_strings_count_fn) {
+		DP_DEBUG(DP_DBG_FLAG_MIB,
+			 "MIB call dp_get_dev_stat_strings_count_fn callback\n");
+		return dp_get_dev_stat_strings_count_fn(dev);
+	}
 
-	return pinfo->dp_net_dev_get_ss_stat_strings_count(dev);
+	return 0;
 }
 EXPORT_SYMBOL(dp_net_dev_get_ss_stat_strings_count);
 
 void dp_net_dev_get_ethtool_stats(struct net_device *dev,
 				  struct ethtool_stats *stats, u64 *data)
 {
-	dp_subif_t subif = {0};
-	struct inst_info *pinfo = &dp_port_prop[subif.inst].info;
-
 	/* This is for the Mib counter wraparound module */
 	if (dp_dev_get_ethtool_stats_fn) {
 		DP_DEBUG(DP_DBG_FLAG_MIB,
@@ -2724,12 +2744,7 @@ void dp_net_dev_get_ethtool_stats(struct net_device *dev,
 		return dp_dev_get_ethtool_stats_fn(dev, stats, data);
 	}
 
-	if (!pinfo->dp_net_dev_get_ethtool_stats)
-		return;
-
-	DP_DEBUG(DP_DBG_FLAG_MIB,
-		 "MIB call pinfo->dp_net_dev_get_ethtool_stats\n");
-	pinfo->dp_net_dev_get_ethtool_stats(dev, stats, data);
+	return;
 }
 EXPORT_SYMBOL(dp_net_dev_get_ethtool_stats);
 
diff --git a/drivers/net/datapath/dpm/datapath_stats.c b/drivers/net/datapath/dpm/datapath_stats.c
deleted file mode 100644
index 2901177148ee..000000000000
--- a/drivers/net/datapath/dpm/datapath_stats.c
+++ /dev/null
@@ -1,28 +0,0 @@
-/*
- * Copyright (C) Intel Corporation
- * Author: Robert Dabrowski <robertx.dabrowski@intel.com>
- *
- * This program is free software; you can redistribute it and/or modify it
- * under the terms of the GNU General Public License version 2 as published
- * by the Free Software Foundation.
- */
-
-#include "datapath.h"
-
-/* Ethtool statistics strings */
-static const char dp_ethtool_gstrings[][ETH_GSTRING_LEN] = {
-#define _DP_ETHTOOL(x, y)	# x,
-	DP_ETHTOOL_STAT_REG_DECLARE
-#undef _DP_ETHTOOL
-};
-
-/* ethtool statistics support */
-int dp_net_dev_get_ss_stat_strings_count_3X(struct net_device *dev)
-{
-	return ARRAY_SIZE(dp_ethtool_gstrings);
-}
-
-void dp_net_dev_get_ss_stat_strings_3X(struct net_device *dev, u8 *data)
-{
-	memcpy(data, dp_ethtool_gstrings, sizeof(dp_ethtool_gstrings));
-}
diff --git a/drivers/net/datapath/dpm/datapath_stats.h b/drivers/net/datapath/dpm/datapath_stats.h
deleted file mode 100644
index d0208b439f31..000000000000
--- a/drivers/net/datapath/dpm/datapath_stats.h
+++ /dev/null
@@ -1,30 +0,0 @@
-/*
- * Copyright (C) Intel Corporation
- * Author: Robert Dabrowski <robertx.dabrowski@intel.com>
- *
- * This program is free software; you can redistribute it and/or modify it
- * under the terms of the GNU General Public License version 2 as published
- * by the Free Software Foundation.
- */
-
-#ifndef DATAPATH_STATS_H
-#define DATAPATH_STATS_H
-
-/*!
- *@brief dp_net_dev_get_ss_stat_strings_count_3X - function used by callback
- *dp_net_dev_get_ss_stat_strings_count to retrieve number of counters
- *@param[in] dev: net device
- *@return number of counters
- */
-int dp_net_dev_get_ss_stat_strings_count_3X(struct net_device *dev);
-
-/*!
- *@brief dp_net_dev_get_ss_stat_strings_3X - function used by gswip callback
- *dp_net_dev_get_ss_stat_strings to retrieve set of counters' names.
- *NOTE: The set of counters names is common for gswip31 and gdwip32
- *@param[in] dev: net device
- *@param[out] data: buffer pointer where the function will copy counters' names
- */
-void dp_net_dev_get_ss_stat_strings_3X(struct net_device *dev, u8 *data);
-
-#endif
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index a07db007923d..15021bee98c4 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -12,7 +12,6 @@
 #include <linux/if_vlan.h>
 #include "../datapath.h"
 #include "../datapath_instance.h"
-#include "../datapath_stats.h"
 #include "datapath_ppv4.h"
 #include "datapath_misc.h"
 
@@ -2382,24 +2381,11 @@ int register_dp_cap_gswip31(int flag)
 	cap.info.dp_set_bp_attr = dp_set_bp_attr_31;
 	cap.info.dp_rx = dp_rx_31;
 	cap.info.dp_tx = dp_xmit_31;
-	cap.info.dp_net_dev_get_ss_stat_strings =
-		dp_net_dev_get_ss_stat_strings_3X;
-	cap.info.dp_net_dev_get_ss_stat_strings_count =
-		dp_net_dev_get_ss_stat_strings_count_3X;
-#if IS_ENABLED(CONFIG_INTEL_MIB_COUNTERS)
 	/* The full ethtool statistics support is provided in mib_counters
 	 * kernel module where special callback fn is registered for the
 	 * statistics purpose
 	 */
 	cap.info.dp_net_dev_get_ethtool_stats = NULL;
-#else
-	/* In case mib_counters module is not compiled in
-	 * here is registered simple counters retrieval function
-	 * without overflow handling
-	 */
-	cap.info.dp_net_dev_get_ethtool_stats =
-		dp_net_dev_get_ethtool_stats_31;
-#endif
 
 	cap.info.dp_alloc_bridge_port = alloc_bridge_port;
 	cap.info.dp_free_bridge_port = free_bridge_port;
@@ -2450,100 +2436,3 @@ int register_dp_cap_gswip31(int flag)
 
 	return 0;
 }
-
-/********************* ETHTOOL STATISTIC SUPPORT ****************************/
-
-#if IS_ENABLED(CONFIG_INTEL_MIB_COUNTERS)
-/* ethtool statistics serviced in mib_counters module */
-#else
-/* This function is used for retrieving MIB counters statistics from
- * HW registers when "mib_counters" kernel module isn't used for
- * handling counters overflow
- */
-void dp_net_dev_get_ethtool_stats_31(struct net_device *dev,
-				     struct ethtool_stats *stats, u64 *data)
-{
-	GSW_RMON_Port_cnt_t *g_counters;
-	struct pmac_port_info *port_info;
-	struct core_ops *ops;
-	dp_subif_t *ps_subif;
-	u32 flags = 0;
-	u64 *data_dst;
-	s32 ret = 0;
-	u8 i = 0;
-
-	ops = dp_port_prop[0].ops[0];
-
-	ps_subif = devm_kzalloc(&g_dp_dev->dev, (sizeof(dp_subif_t)),
-				GFP_ATOMIC);
-	if (!ps_subif)
-		return;
-
-	/* two counters structures are used
-	 * one set for BP or regular CTP
-	 * second set for CTP PCE-Bypass
-	 */
-	g_counters = devm_kzalloc(&g_dp_dev->dev,
-				  (2 * sizeof(GSW_RMON_Port_cnt_t)),
-				  GFP_ATOMIC);
-	if (!g_counters) {
-		devm_kfree(&g_dp_dev->dev, ps_subif);
-		return;
-	}
-	/* get subif information from net device */
-	ret = dp_get_netif_subifid(dev, NULL, NULL, NULL, ps_subif, flags);
-	if (ret != DP_SUCCESS) {
-		netdev_err(dev, "dp_get_netif_subifid returned %d\n", ret);
-		devm_kfree(&g_dp_dev->dev, ps_subif);
-		devm_kfree(&g_dp_dev->dev, g_counters);
-		return;
-	}
-
-	/* for all port kinds use the same algorithm:
-	 * read Rx counters using CTP_PORT (even for BRIDGE ports)
-	 * read Tx counters using CTP Bypass
-	 */
-	port_info = get_dp_port_info(0, ps_subif->port_id);
-	/* first pass reads CTP counters,
-	 * second pass reads CTP-Bypass counters
-	 */
-	for (i = 0; i < 2; i++) {
-		g_counters[i].ePortType = GSW_CTP_PORT;
-		g_counters[i].nPortId = ps_subif->port_id;
-		g_counters[i].nSubIfIdGroup = GET_VAP(ps_subif->subif,
-						      port_info->vap_offset,
-						      port_info->vap_mask);
-		g_counters[i].bPceBypass = i;
-
-		/* read CTP counters */
-		ret = ops->gsw_rmon_ops.RMON_Port_Get(ops, &g_counters[i]);
-		if (ret != DP_SUCCESS) {
-			netdev_err(dev, "RMON_Port_Get returned %d\n", ret);
-			devm_kfree(&g_dp_dev->dev, g_counters);
-			devm_kfree(&g_dp_dev->dev, ps_subif);
-			return;
-		}
-	}
-
-	/* CTP TX counters are placed before QoS modules, and they
-	 * record packets dropped by QoS modules too.
-	 * CTP TX Bypass counters are placed after QoS modules,
-	 * and they record the packets actually transmitted.
-	 */
-#define _DP_ETHTOOL_TX(r)	g_counters[0].r = g_counters[1].r;
-	DP_ETHTOOL_STAT_REG_TX_DECLARE
-#undef _DP_ETHTOOL_TX
-
-	/* here subtract counters by reinserted packets */
-
-	/* copy counters' values to ethtool buffer */
-	data_dst = data;
-
-#define _DP_ETHTOOL(x, y)	*data_dst++ = g_counters[0].y;
-	DP_ETHTOOL_STAT_REG_DECLARE
-#undef _DP_ETHTOOL
-
-	devm_kfree(&g_dp_dev->dev, g_counters);
-	devm_kfree(&g_dp_dev->dev, ps_subif);
-}
-#endif /* CONFIG_INTEL_MIB_COUNTERS */
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.h b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
index 3a68baae809d..9b928cf4e0c6 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.h
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
@@ -306,19 +306,4 @@ int tc_vlan_set_31(struct core_ops *ops, struct dp_tc_vlan *vlan,
 int dp_get_reinsert_cnt_31(int inst, int dp_port, int vap, int flag,
 			   struct dp_reinsert_count *dp_reins_count);
 
-#if IS_ENABLED(CONFIG_INTEL_MIB_COUNTERS)
-/* ethtool statistics serviced in mib_counters module */
-#else
-/*!
- *@brief dp_net_dev_get_ethtool_stats_31 - function used by callback
- * dp_net_dev_get_ethtool_stats to retrieve counters' values specyfic for
- * gswip31
- *@param[in] dev: net device
- *@param[out] stats: for dumping NIC-specific statistics
- *@param[out] data: for dumping counters values sequence
- */
-void dp_net_dev_get_ethtool_stats_31(struct net_device *dev,
-				     struct ethtool_stats *stats, u64 *data);
-#endif /* CONFIG_INTEL_MIB_COUNTERS */
-
 #endif /* DATAPATH_MISC31_H */
diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index d3c5307d6802..be2b68b6d9db 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -27,12 +27,69 @@
 
 #define INITIAL_UPDATE 1
 
+/* Ethtool strings and counter names (based on GSW_RMON_Port_cnt_t) */
+#define DP_GSW_ETHTOOL_STAT_REG_DECLARE	\
+	_DP_ETHTOOL(RxExtendedVlanDiscardPkts, nRxExtendedVlanDiscardPkts) \
+	_DP_ETHTOOL(MtuExceedDiscardPkts, nMtuExceedDiscardPkts) \
+	_DP_ETHTOOL(TxUnderSizeGoodPkts, nTxUnderSizeGoodPkts) \
+	_DP_ETHTOOL(TxOversizeGoodPkts, nTxOversizeGoodPkts) \
+	_DP_ETHTOOL(RxGoodPkts, nRxGoodPkts) \
+	_DP_ETHTOOL(RxUnicastPkts, nRxUnicastPkts) \
+	_DP_ETHTOOL(RxBroadcastPkts, nRxBroadcastPkts) \
+	_DP_ETHTOOL(RxMulticastPkts, nRxMulticastPkts) \
+	_DP_ETHTOOL(RxFCSErrorPkts, nRxFCSErrorPkts) \
+	_DP_ETHTOOL(RxUnderSizeGoodPkts, nRxUnderSizeGoodPkts) \
+	_DP_ETHTOOL(RxOversizeGoodPkts, nRxOversizeGoodPkts) \
+	_DP_ETHTOOL(RxUnderSizeErrorPkts, nRxUnderSizeErrorPkts) \
+	_DP_ETHTOOL(RxGoodPausePkts, nRxGoodPausePkts) \
+	_DP_ETHTOOL(RxOversizeErrorPkts, nRxOversizeErrorPkts) \
+	_DP_ETHTOOL(RxAlignErrorPkts, nRxAlignErrorPkts) \
+	_DP_ETHTOOL(RxFilteredPkts, nRxFilteredPkts) \
+	_DP_ETHTOOL(Rx64BytePkts, nRx64BytePkts) \
+	_DP_ETHTOOL(Rx127BytePkts, nRx127BytePkts) \
+	_DP_ETHTOOL(Rx255BytePkts, nRx255BytePkts) \
+	_DP_ETHTOOL(Rx511BytePkts, nRx511BytePkts) \
+	_DP_ETHTOOL(Rx1023BytePkts, nRx1023BytePkts) \
+	_DP_ETHTOOL(RxMaxBytePkts, nRxMaxBytePkts) \
+	_DP_ETHTOOL(TxGoodPkts, nTxGoodPkts) \
+	_DP_ETHTOOL(TxUnicastPkts, nTxUnicastPkts) \
+	_DP_ETHTOOL(TxBroadcastPkts, nTxBroadcastPkts) \
+	_DP_ETHTOOL(TxMulticastPkts, nTxMulticastPkts) \
+	_DP_ETHTOOL(TxSingleCollCount, nTxSingleCollCount) \
+	_DP_ETHTOOL(TxMultCollCount, nTxMultCollCount) \
+	_DP_ETHTOOL(TxLateCollCount, nTxLateCollCount) \
+	_DP_ETHTOOL(TxExcessCollCount, nTxExcessCollCount) \
+	_DP_ETHTOOL(TxCollCount, nTxCollCount) \
+	_DP_ETHTOOL(TxPauseCount, nTxPauseCount) \
+	_DP_ETHTOOL(Tx64BytePkts, nTx64BytePkts) \
+	_DP_ETHTOOL(Tx127BytePkts, nTx127BytePkts) \
+	_DP_ETHTOOL(Tx255BytePkts, nTx255BytePkts) \
+	_DP_ETHTOOL(Tx511BytePkts, nTx511BytePkts) \
+	_DP_ETHTOOL(Tx1023BytePkts, nTx1023BytePkts) \
+	_DP_ETHTOOL(TxMaxBytePkts, nTxMaxBytePkts) \
+	_DP_ETHTOOL(TxDroppedPkts, nTxDroppedPkts) \
+	_DP_ETHTOOL(TxAcmDroppedPkts, nTxAcmDroppedPkts) \
+	_DP_ETHTOOL(RxDroppedPkts, nRxDroppedPkts) \
+	_DP_ETHTOOL(RxGoodBytes, nRxGoodBytes) \
+	_DP_ETHTOOL(RxBadBytes, nRxBadBytes) \
+	_DP_ETHTOOL(TxGoodBytes, nTxGoodBytes)
+
+/* Ethtool strings and counter names for all counters */
+#define DP_ETHTOOL_STAT_REG_DECLARE \
+	DP_GSW_ETHTOOL_STAT_REG_DECLARE \
+	_DP_ETHTOOL(RxOverflowError, nRxOverflowError)
+
 struct mib_counters_u64 {
 #define _DP_ETHTOOL(x, y)	u64 y;
 DP_ETHTOOL_STAT_REG_DECLARE
 #undef _DP_ETHTOOL
 };
 
+struct last_counters {
+	GSW_RMON_Port_cnt_t rmon;
+	u64 rx_ovfl_err;
+};
+
 /** Data set needed for maintain mib counters overflow for one net device */
 struct mib_counters {
 	/** devices list handle */
@@ -50,11 +107,18 @@ struct mib_counters {
 	/** flag indicates dev is UNI side bridge port */
 	u8 dev_is_uni;
 	/** last read counters */
-	GSW_RMON_Port_cnt_t last_read_rmon;
+	struct last_counters last;
 	/** stored 64 bits wide counters */
 	struct mib_counters_u64 wide_cntrs;
 };
 
+/* Ethtool statistics strings */
+static const char dp_ethtool_gstrings[][ETH_GSTRING_LEN] = {
+#define _DP_ETHTOOL(x, y)	# x,
+	DP_ETHTOOL_STAT_REG_DECLARE
+#undef _DP_ETHTOOL
+};
+
 /* Timer for update counters for all registered net devices */
 static struct timer_list cnt_upd_timer;
 
@@ -85,16 +149,14 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 {
 	GSW_RMON_Port_cnt_t *tmp_counters;
 	struct core_ops *ops;
+	struct mac_ops *mac_ops;
+	struct mac_rmon *rmon = NULL;
 	s32 ret = 0;
 	u8 i = 0;
 
 	tmp_counters = devm_kzalloc(&g_dp_dev->dev,
 				    (2 * sizeof(GSW_RMON_Port_cnt_t)),
 				    GFP_ATOMIC);
-	if (!tmp_counters) {
-		netdev_err(cnt->dev, "failed to get memory for counters");
-		return;
-	}
 
 	ops = dp_port_prop[cnt->inst].ops[0];
 
@@ -151,6 +213,28 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 #undef _DP_ETHTOOL_TX
 	}
 
+	/* mac_ops is optional, do not fail if unavailable */
+	mac_ops = dp_port_prop[cnt->inst].mac_ops[cnt->port_id];
+	if (mac_ops) {
+		rmon = devm_kzalloc(&g_dp_dev->dev, sizeof(struct mac_rmon),
+				    GFP_ATOMIC);
+		if (rmon) {
+			mac_ops->rmon_get(mac_ops, rmon);
+			tmp_counters[0].nRxGoodPausePkts =
+				rmon->cnt[RMON_RX_PAUSE];
+			tmp_counters[0].nTxPauseCount =
+				rmon->cnt[RMON_TX_PAUSE];
+			tmp_counters[0].nTxExcessCollCount =
+				rmon->cnt[RMON_TX_EXCS_COL];
+			tmp_counters[0].nTxLateCollCount =
+				rmon->cnt[RMON_TX_LATE_COL];
+			tmp_counters[0].nTxSingleCollCount =
+				rmon->cnt[RMON_TX_SINGLE_COL];
+			tmp_counters[0].nTxMultCollCount =
+				rmon->cnt[RMON_TX_MULTI_COL];
+		}
+	}
+
 	/* On initial update, just after new device is added to update
 	 * counter list, counters values read first time can be invalid.
 	 * Therefore the value just read will not be added to the 64 bits
@@ -158,22 +242,34 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 	 */
 	if (initial_update == INITIAL_UPDATE) {
 		write_lock(&cnt->cnt_lock);
-		cnt->last_read_rmon = tmp_counters[0];
+		cnt->last.rmon = tmp_counters[0];
+		if (rmon)
+			cnt->last.rx_ovfl_err =	rmon->cnt[RMON_RX_OVERFLOW];
 		write_unlock(&cnt->cnt_lock);
 		return;
 	}
 
 	/* update 64 bits wide counters based on last read values */
 	write_lock(&cnt->cnt_lock);
-#define _DP_ETHTOOL(x, y) cnt->wide_cntrs.y += wraparound(tmp_counters[0].y, cnt->last_read_rmon.y, sizeof(cnt->last_read_rmon.y));
-	DP_ETHTOOL_STAT_REG_DECLARE
+#define _DP_ETHTOOL(x, y) cnt->wide_cntrs.y += wraparound(tmp_counters[0].y, cnt->last.rmon.y, sizeof(tmp_counters[0].y));
+	DP_GSW_ETHTOOL_STAT_REG_DECLARE
 #undef _DP_ETHTOOL
 
+	if (rmon)
+		cnt->wide_cntrs.nRxOverflowError +=
+			wraparound(rmon->cnt[RMON_RX_OVERFLOW],
+				   cnt->last.rx_ovfl_err,
+				   sizeof(rmon->cnt[RMON_RX_OVERFLOW]));
+
 	/* Update last read counters set */
-	cnt->last_read_rmon = tmp_counters[0];
+	cnt->last.rmon = tmp_counters[0];
+	if (rmon)
+		cnt->last.rx_ovfl_err =	rmon->cnt[RMON_RX_OVERFLOW];
 	write_unlock(&cnt->cnt_lock);
 
 	devm_kfree(&g_dp_dev->dev, tmp_counters);
+	if (rmon)
+		devm_kfree(&g_dp_dev->dev, rmon);
 }
 
 /* update counters for all registered devices */
@@ -461,6 +557,16 @@ void mib_cnt_get_ethtool_stats_31(struct net_device *dev,
 #undef _DP_ETHTOOL
 }
 
+static int dp_dev_get_stat_strings_count(struct net_device *dev)
+{
+	return ARRAY_SIZE(dp_ethtool_gstrings);
+}
+
+static void dp_dev_get_stat_strings(struct net_device *dev, u8 *data)
+{
+	memcpy(data, dp_ethtool_gstrings, sizeof(dp_ethtool_gstrings));
+}
+
 /********************* TIMER **************************************/
 
 void cnt_upd_timer_handler(unsigned long data)
@@ -512,6 +618,8 @@ static int __init mib_counters_mod_init(void)
 
 	/* Register ethtool stats cb function */
 	dp_set_ethtool_stats_fn(0, mib_cnt_get_ethtool_stats_31);
+	dp_set_ethtool_stats_strings_cnt_fn(0, dp_dev_get_stat_strings_count);
+	dp_set_ethtool_stats_strings_fn(0, dp_dev_get_stat_strings);
 
 	return 0;
 }
@@ -551,6 +659,8 @@ static void __exit mib_counters_mod_exit(void)
 
 	/* Deregister ethtool stats cb function */
 	dp_set_ethtool_stats_fn(0, NULL);
+	dp_set_ethtool_stats_strings_cnt_fn(0, NULL);
+	dp_set_ethtool_stats_strings_fn(0, NULL);
 }
 
 module_init(mib_counters_mod_init);
diff --git a/include/net/datapath_api.h b/include/net/datapath_api.h
index 7d22e7006e3c..f1606c238273 100644
--- a/include/net/datapath_api.h
+++ b/include/net/datapath_api.h
@@ -1982,7 +1982,7 @@ int dp_register_event_cb(struct dp_event *info, uint32_t flag);
  *@brief dp_set_ethtool_stats_fn API to register callback function which
  * retrieves counters for ethtool statistics
  *@param[in] inst: dp instance id
- *@param[in] cb: pointer to function called for retrieve counters for
+ *@param[in] cb: pointer to function called to retrieve counters for
  *		 ethtool statistics
  */
 void dp_set_ethtool_stats_fn(int inst,
@@ -1990,6 +1990,26 @@ void dp_set_ethtool_stats_fn(int inst,
 					struct ethtool_stats *stats,
 					u64 *data));
 
+/*!
+ *@brief dp_set_ethtool_stats_strings_cnt_fn API to register callback function which
+ * retrieves retrieve number of counters
+ *@param[in] inst: dp instance id
+ *@param[in] cb: pointer to function called to retrieve number of counters for
+ *		 ethtool statistics
+ */
+void dp_set_ethtool_stats_strings_cnt_fn(int inst,
+					 int (*cb)(struct net_device *dev));
+
+/*!
+ *@brief dp_set_ethtool_stats_strings_fn API to register callback function which
+ * retrieves set of counters' names
+ *@param[in] inst: dp instance id
+ *@param[in] cb: pointer to function called to retrieve set of counters' names
+ *		 for ethtool statistics
+ */
+void dp_set_ethtool_stats_strings_fn(int inst,
+				     void (*cb)(struct net_device *dev, u8 *data));
+
 /*! @brief Enumerator DP_OPS_TYPE */
 enum DP_OPS_TYPE {
 	DP_OPS_BM, /*!< BM ops type */
